src/Angular.js

```diff
@@ -740,6 +740,49 @@ function arrayRemove(array, value) {
   return index;
 }
 
+
+function ES6MapShim() {
+  this._keys = [];
+  this._values = [];
+}
+ES6MapShim.prototype = {
+  get: function(key) {
+    var idx = this._keys.indexOf(key);
+    if (idx !== -1) {
+      return this._values[idx];
+    }
+  },
+  set: function(key, value) {
+    var idx = this._keys.indexOf(key);
+    if (idx === -1) {
+      idx = this._keys.length;
+    }
+    this._keys[idx] = key;
+    this._values[idx] = value;
+    return this;
+  }
+};
+
+function testES6Map(Map) {
+  var m, o = {};
+  return isFunction(Map) && (m = new Map())
+
+    // Required functions
+    && isFunction(m.get) && isFunction(m.set)
+
+    // Number keys, must not call toString
+    && m.get(1) === undefined
+    && m.set(1, o) === m
+    && m.get(1) === o
+    && m.get("1") === undefined
+
+    // Object keys, must use instance as key and not call toString
+    && m.set(o, 2) === m && m.get(o) === 2
+    && m.get({}) === undefined;
```

Comment:
Might not be necessary. I was more concerned about bad non-native shims / helpers. An is-native check might also work but I don't think that is done anywhere else...